metadata_mapper/mappers/validator.py

```diff
@@ -0,0 +1,740 @@
+import itertools
+
+from datetime import datetime
+from typing import Any, Callable, IO, Union
+
+import utilities
+
+
+class ValidationErrors:
+    CSV_FIELDS: dict[str, str] = {
+        "harvest_id": "Harvest ID",
+        "level": "Level",
+        "field": "Field",
+        "description": "Description",
+        "expected": "Expected Value",
+        "actual": "Actual Value"
+    }
+
+    def __init__(self):
+        self.errors: list[dict[str, Any]] = []
+
+    def has_errors(self):
+        return True if len(self.errors) else False
+
+    def is_empty(self):
+        return not self.has_errors()
+
+    def add(self, key: str, field: str, description: str, expected: Any,
+            actual: Any, level: str = "ERROR", **context) -> None:
+        """
+        Adds an error to the errors list.
+
+        Parameters:
+            key: str
+                The harvest ID (or other key) used to join the Rikolti
+                and comparison record.
+            field: str
+                The field on which the error occurred
+            description: str
+                Error description
+            expected: Any
+                Expected value (probably from Solr)
+            actual: Any
+                Actual value (from Rikolti mapper)
+            type: str
+                Error type (ERROR, WARN, etc)
+            context: dict
+                Additional values to be added (will not have headers)
+        """
+        self.errors.append({
+            "harvest_id": key,
+            "level": level,
+            "description": description,
+            "field": field,
+            "expected": str(expected),
+            "actual": str(actual),
+            **context
+        })
+
+    def merge(self, other_errors: "ValidationErrors") -> "ValidationErrors":
+        """
+        Merges errors list with another.
+
+        Parameters:
+            other_errors: ValidationErrors
+                Another ValidationErrors instance to merge in
+
+        Returns: ValidationErrors
+        """
+        if not other_errors.errors:
+            return
+
+        self.errors = self.errors + other_errors.errors
+        return self.errors
+
+    def output(self, format: str, file: IO[str] = None,
+               **opts) -> list[Union[list, dict]]:
+        """
+        Generic method to output data.
+
+        TODO: Expand types of output (i.e. to console for testing)
+
+        Parameters:
+            format: str
+                "stdout" or "csv"
+            file: IO[str] (default: None)
+                A file name/path to write to (for CSV output)
+        """
+        if format == "stdout":
+            return print(self.errors)
+        elif format == "csv" and file:
+            return self.output_csv(file)
```

Comment:
I've used `output` mainly for dev, to get errors printed to the console. It is not currently used in the `validate_mapping` workflow.

The original idea was to have a single interface to output, and to use an argument (`format`) to determine how it gets output. That way, different workflows could output however they'd like. However, I realized that this didn't really work for existing S3 output (which writes using boto3 - this output method requires a file reference), and instead used the existing logic (now in `utilities`) to write to the env-appropriate bucket.

We could remove `output`, but I think it would still be nice to have a convenience function to output to stdout. Perhaps something simply named `output_to_stdout`?